xmtp_mls/src/groups/mls_sync.rs

```diff
@@ -1184,95 +1224,89 @@ where
+                            hex::encode(envelope.group_id.as_slice()),
+                            cursor
+                        );
+                        let current_cursor = provider
+                            .db()
+                            .get_last_cursor_for_id(&envelope.group_id, EntityKind::Group)?;
+                        current_cursor < cursor as i64
+                    };
+                    if !requires_processing {
+                        tracing::debug!("message @cursor=[{}] for group=[{}] created_at=[{}] no longer require processing, should be available in database",
+                            envelope.id,
+                            xmtp_common::fmt::debug_hex(&envelope.group_id),
+                            envelope.created_ns
+                        );
 
-                            // early return if the message is already procesed
-                            // _NOTE_: Not early returning and re-processing a message that
-                            // has already been processed, has the potential to result in forks.
-                            // In some cases, we may want to roll back the cursor if we updated the
-                            // cursor, but actually cannot process the message.
-                            identifier.previously_processed(true);
-                            return Ok(());
-                        }
-                        let (intent_state, internal_message_id) = match maybe_validated_commit {
-                            // the error is also a Result
-                            Err(err) => match err {
-                                Ok(intent_state) => (intent_state, None),
-                                Err(e) => return Err(e)
-                            },
-                            Ok(commit) => {
-                                self
-                                .process_own_message(&mut mls_group, commit, &intent, &message, envelope)?
+                        // early return if the message is already procesed
+                        // _NOTE_: Not early returning and re-processing a message that
+                        // has already been processed, has the potential to result in forks.
+                        // In some cases, we may want to roll back the cursor if we updated the
+                        // cursor, but actually cannot process the message.
+                        identifier.previously_processed(true);
+                        return Ok(());
+                    }
+                    let (intent_state, internal_message_id) = match maybe_validated_commit {
+                        Err(err) => {
+                            if err.processing_error.is_retryable() {
+                                // Rollback the transaction so that we can retry
+                                return Err(err.processing_error);
                             }
```

Comment:
We are processing the error from above here. I don't think `stage_and_validate_intent()` returns any retriable errors, because there are no DB or networking calls, so I don't think this is a functional change.

I chose to include this logic anyway, in case `stage_and_validate_intent()` is modified to return a retriable error in the future - in that case, we should just return an error and rollback everything.

Reply:
These are great finds